File: server/imgworker.js
/* THE WARM PYTHON, AND WHY THERE IS ONE.
 *
 * Measured on this machine at 1024x1024, through the live route:
 *
 *     one brush stroke, end to end         653 ms
 *     the same call with ops={} - NO WORK   596 ms
 *     the stroke rasterisation itself        11 ms
 *
 * An edit that does nothing costs 91% of an edit that does something, because
 * every request spawns a fresh interpreter and pays numpy + cv2 + PIL again
 * (~390 ms of it). Through this worker the same edits measure 126 ms for the
 * first and then 65-71 ms - about ten times faster, and inside the budget where
 * dragging a brush stops feeling like submitting a form.
 *
 * ⚠ ONE WORKER, SERIALISED, AND THAT IS DELIBERATE. The engine is numpy on a
 * single picture; two concurrent renders would contend for the same cores and
 * the same cache, and a pool would need every caller to think about which
 * worker holds which decoded source. Requests queue in arrival order.
 *
 * ⚠ IT IS FOR PREVIEWS, NOT FOR COMMITS. Nothing here writes the library, files
 * provenance or makes a thumbnail. The committing routes keep their own spawn:
 * a preview that dies costs a frame, and a commit that dies costs the work.
 */
import { spawn } from "node:child_process";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

export class ImgWorker {
  constructor(python, { onLog = () => {} } = {}) {
    this.python = python;
    this.onLog = onLog;
    this.proc = null;
    this.buf = "";
    this.next = 1;
    this.pending = new Map();
    this.ready = null;
  }

  /* Spawned on first use rather than at boot: a studio that never opens the
   * image editor should not hold an idle interpreter, and the 258 ms handshake
   * is paid once by whoever asks first. */
  start() {
    if (this.ready) return this.ready;
    this.ready = new Promise((resolve, reject) => {
      const proc = spawn(this.python, [path.join(__dirname, "imgworker.py")], { windowsHide: true });
      this.proc = proc;
      let settled = false;
      proc.stdout.on("data", (d) => {
        this.buf += d;
        let i;
        while ((i = this.buf.indexOf("\n")) >= 0) {
          const line = this.buf.slice(0, i).trim();
          this.buf = this.buf.slice(i + 1);
          if (!line) continue;
          let msg;
          try { msg = JSON.parse(line); } catch { this.onLog(`[imgworker] ${line.slice(0, 200)}`); continue; }
          if (msg.ready && !settled) { settled = true; resolve(this); continue; }
          const w = this.pending.get(msg.id);
          if (w) { this.pending.delete(msg.id); w.resolve(msg); }
        }
      });
      proc.stderr.on("data", (d) => this.onLog(`[imgworker] ${String(d).trim().slice(0, 400)}`));
      /* ⚠ A DEAD WORKER MUST REJECT EVERY WAITER. Without this the caller's
       * promise never settles and the request hangs until the browser gives
       * up — the failure mode that looks like the studio freezing. */
      const die = (why) => {
        this.proc = null; this.ready = null;
        for (const [, w] of this.pending) w.reject(new Error(`image worker ${why}`));
        this.pending.clear();
        if (!settled) { settled = true; reject(new Error(`image worker ${why}`)); }
      };
      proc.on("error", (err) => die(`could not start: ${err.message}`));
      proc.on("exit", (code) => die(`exited (${code})`));
    });
    return this.ready;
  }

  /** One job. Resolves with the engine's own reply object. */
  async run(mode, job, { timeoutMs = 30_000 } = {}) {
    await this.start();
    const id = this.next++;
    const reply = await new Promise((resolve, reject) => {
      this.pending.set(id, { resolve, reject });
      const t = setTimeout(() => {
        if (this.pending.delete(id)) reject(new Error(`image worker timed out after ${timeoutMs} ms`));
      }, timeoutMs);
      const done = (fn) => (v) => { clearTimeout(t); fn(v); };
      this.pending.set(id, { resolve: done(resolve), reject: done(reject) });
      try {
        this.proc.stdin.write(`${JSON.stringify({ id, mode, job })}\n`);
      } catch (err) {
        this.pending.delete(id); clearTimeout(t); reject(err);
      }
    });
    if (reply.ok === false) throw new Error(reply.error || "image worker refused the job");
    return reply;
  }

  stop() {
    if (this.proc) { try { this.proc.stdin.end(); } catch { /* already gone */ } }
    this.proc = null; this.ready = null;
  }
}

File: server/imgworker.py
"""A PYTHON THAT STAYS, BECAUSE SPAWNING ONE IS 60% OF EVERY EDIT.

Measured on this machine, at 1024x1024, through the live route:

    one brush stroke, end to end            653 ms
    the same call with ops={} - NO WORK     596 ms
    bare interpreter start                  ~160 ms
    + numpy, cv2, PIL imported              ~390 ms
    the stroke rasterisation itself          11 ms

An edit that does nothing costs 91% of an edit that does something, because
server/index.js spawns a FRESH interpreter for every request and pays the import
cost each time. The pixels were never the problem.

This module is the same engine behind a process that does not exit. One request
per line of stdin, one reply per line of stdout:

    -> {"id": 7, "mode": "edit", "job": {...}}
    <- {"id": 7, "ok": true, "out": "...", "width": 1024, "height": 768}

⚠ apply_edit PRINTS ITS RESULT, so the protocol has to take stdout away from
it. imagetools is a CLI first: apply_edit ends by printing a JSON line, which
is exactly what this file uses stdout for. Left alone, the engine's own report
would interleave with the protocol and every reply would be unparseable. So each
call runs inside redirect_stdout and the captured line becomes the reply's body
- the engine keeps its contract, and the pipe stays clean.

⚠ AND A CRASH MUST ANSWER, NOT DIE. A worker that exits on a bad job takes
every queued request with it and the node side waits forever. Every exception
becomes a reply with the traceback's last line, and the loop continues; only a
closed stdin ends it.
"""
import contextlib
import io
import json
import os
import sys
import traceback

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

import imagetools  # noqa: E402


def _call(mode, job):
    """Run one job with stdout borrowed, and give back what it printed."""
    buf = io.StringIO()
    with contextlib.redirect_stdout(buf):
        if mode == "edit":
            imagetools.apply_edit(job)
        elif mode == "blank":
            imagetools.blank(job)
        elif mode == "describe":
            imagetools.describe_selection(job)
        elif mode == "composite":
            imagetools.composite(job)
        else:
            raise ValueError(f"unknown mode {mode!r}")
    said = buf.getvalue().strip()
    if not said:
        return {}
    # The engine's own report is the LAST line; anything before it is a note it
    # printed on the way (imgstroke and imgselect both warn on stdout).
    try:
        return json.loads(said.splitlines()[-1])
    except Exception:                                   # noqa: BLE001
        return {"said": said[-400:]}


def main():
    # The handshake, so node can wait for the imports rather than guessing.
    sys.stdout.write(json.dumps({"ready": True, "pid": os.getpid()}) + "\n")
    sys.stdout.flush()

    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        rid = None
        try:
            req = json.loads(line)
            rid = req.get("id")
            body = _call(req.get("mode") or "edit", req.get("job") or {})
            reply = {"id": rid, "ok": True, **body}
        except Exception as exc:                        # noqa: BLE001
            # A worker that dies on a bad job strands every queued request.
            tb = traceback.format_exc().strip().splitlines()
            reply = {"id": rid, "ok": False,
                     "error": f"{type(exc).__name__}: {exc}",
                     "where": tb[-3] if len(tb) >= 3 else ""}
        sys.stdout.write(json.dumps(reply) + "\n")
        sys.stdout.flush()


if __name__ == "__main__":
    main()
